scientific-computing-with-python/learn-object-oriented-programming-by-building-a-coffee-making-machine/CoffeeMachine.py

```diff
@@ -0,0 +1,153 @@
+from time import sleep
+
+"""
+The Maker class represents the coffee maker itself. It keep tracks of the water, milk,
+and beans in the machine. Makes coffee and reports on the resources.
+"""
+
+
+class Maker:
+
+    def __init__(self):
+        self.resources = {
+            "water": 300,
+            "milk": 200,
+            "beans": 100,
+        }
+
+    def report(self):
+        print(f"Water: {self.resources['water']}ml")
+        print(f"Milk: {self.resources['milk']}ml")
+        print(f"Coffee: {self.resources['beans']}g")
+
+    def is_resource_sufficient(self, ordered_menu_item):
+        can_make = True
+        for item in ordered_menu_item.ingredients:
+            if ordered_menu_item.ingredients[item] > self.resources[item]:
+                print(f"Sorry there is not enough {item}.")
+                can_make = False
+        return can_make
+
+    def make_coffee(self, ordered_menu_item):
+        for item in ordered_menu_item.ingredients:
+            self.resources[item] -= ordered_menu_item.ingredients[item]
+        print(f"Here's your {ordered_menu_item.name}. Enjoy! :)")
+
+
+"""
+The MenuItem class represents a menu item. It keeps track of the name, the ingredients,
+and the cost.
+"""
+
+
+class MenuItem:
+
+    def __init__(self, name, water, milk, beans, cost):
+        self.name = name
+        self.cost = cost
+        self.ingredients = {
+            'water': water,
+            'milk': milk,
+            'beans': beans,
+        }
```

Comment:
I think it would be better to pass ingredients as a list/dictionary, without specifically requiring all of them (ie. espresso already needs `0` water, so including it in ingredients is a bit obsolete).

This could be done in two ways:
```suggestion
    def __init__(self, name, ingredients, cost):
        self.name = name
        self.cost = cost
        self.ingredients = ingredients
```
Alternatively, with **kwargs:
```suggestion
    def __init__(self, name, cost, **ingredients):
        self.name = name
        self.cost = cost
        self.ingredients = ingredients
```

Reply:
I'll need to think a bit more about that. I agree with this being outdated though.